composer/trainer/dist_strategy.py

```diff
@@ -247,9 +250,60 @@ def prepare_fsdp_module(model: torch.nn.Module, optimizers: Optional[Union[torch
     for name, obj in model.named_children():
         if not isinstance(obj, (Metric, MetricCollection)):
 
-            # If `obj` contains meta tensors, try to use `obj.param_init_fn` to initialize them
             def _param_init_fn(module: torch.nn.Module) -> None:
-                module.to_empty(device=f'cuda:{torch.cuda.current_device()}')
+                # A dictionary of all tied parameter pointers to module names
+                tied_pointers = {}
+
+                # Goes through all modules finding which weights have the same pointers
+                for name, mod in module.named_modules():
+                    for attr in ['weight', 'bias']:
+                        if hasattr(mod, attr):
+                            ptr = id(getattr(mod, attr))
+                            ptr_attr = (ptr, attr)
+                            tied_pointers[ptr_attr] = tied_pointers.get(ptr_attr, set()) | set([name])
+
+                # Creates a dictionary of module names should be tied together
+                tied_mod_names = collections.defaultdict(lambda: [])
+                # Creates a set of modules we should not initialize
+                should_not_init_params = set()
+                for ptr_attr_type, s in tied_pointers.items():
+                    _, attr_type = ptr_attr_type
+                    first = next(s.__iter__())
+                    for elem in s:
+                        should_not_init_params.add('.'.join([elem, attr_type]))
+                        tied_mod_names[(first, attr_type)].append(elem)
+                    # Make sure at least one of the tied parameters is initialized
+                    should_not_init_params.remove('.'.join([first, attr_type]))
+
+                meta_safe_apply(module,
+                                lambda t: torch.empty_like(t, device=f'cuda:{torch.cuda.current_device()}'),
```

Comment:
I think we always had assumed it should be a CUDA device? 

https://github.com/mosaicml/composer/blob/dev/composer/trainer/dist_strategy.py#L252
https://github.com/mosaicml/composer/blob/dev/composer/trainer/dist_strategy.py#L283

It could be another 'xla' device or sth though... but I'm not sure if we've tested it?

Reply:
Ah, if it is the same as it was before, that is ok. Could you make a JIRA to check on this though?

Reply:
Yeah done, CO-1622